### ISSUE TITLE
Watcher of date type being triggered when dispatch/trigger

### ISSUE THREAD

**Issue by jackysee** [eddyerburgh/avoriaz#73]
Given a following component (Bar.vue):
```vue
<template>
  <div>
    <div class="date">{{dateValue2.toISOString()}}</div>
    <div class="int">{{intValue2}}</div>
    <button @click="change">Change</button>
  </div>
</template>

<script>
export default {
  props: ['dateValue', 'intValue'],
  data(){
    return {
      dateValue2: this.dateValue,
      intValue2: this.intValue
    };
  },
  watch: {
    dateValue(newVal, oldVal){
      this.dateValue2 = this.dateValue;
    },
    intValue(newVal, oldVal){
      this.intValue2 = this.intValue;
    }
  },
  methods:{
    change(){
      this.dateValue2 = new Date(
        this.dateValue2.getFullYear(),
        this.dateValue2.getMonth(),
        1
      );
      this.intValue2 += 1;
    }
  }
};
</script>
```

And this test
```js
import { mount } from 'avoriaz'
import Bar from '../src/components/Bar.vue'
var test = require('tape')

test('set date', (t) => {
  const wrapper = mount(Bar, {
    propsData:{
      dateValue: new Date(2017, 5, 21),
      intValue: 1
    }
  });
  let date = wrapper.find('.date')[0];
  let int = wrapper.find('.int')[0];
  t.is(date.text(), new Date(2017, 5, 21).toISOString(), 'dateValue2 before');
  t.is(int.text(), '1', 'intValue2 before');

  wrapper.find('button')[0].dispatch('click');

  t.is(date.text(), new Date(2017, 5, 1).toISOString(), 'dateValue2 after');
  t.is(int.text(), '2', 'intValue2 after');

  t.end();
});
```

The test for `dateValue2 after` would fail. I found that the `dispatch('click')` has somehow triggered the watcher `dateValue` thus re-assign the value. 

This test would pass for version `1.14.0`, but fail at current `2.4.0`.

I have created the test here:
https://github.com/jackysee/avoriaz-tape-example

**Comment by jackysee** [comment]
The case is still the same if I switch the dispatch to:
```
wrapper.vm.change();
wrapper.update();
```

**Comment by eddyerburgh** [comment]
Yeah, `update` runs all the watchers, and `trigger` runs update

I'll have to think about this over the weekend 😀

**Comment by jackysee** [comment]
I need to have `shallow` for other case. I now just check the vm values.
```
wrapper.vm.change();
t.is(wrapper.vm.dateValue2.toISOString(), new Date(2017, 5, 21).toISOString(), 'dateValue2 after');
```

**Comment by eddyerburgh** [comment]
Ok, I'm going to fix this issue. 

I'm going to remove running the watchers from update and only use it in `setProps` and `setData`, which is where it is needed. I'll only run watchers for values that have been mutated by `setProps` and `setData`.

**Comment by eddyerburgh** [comment]
This should be fixed in 2.4.1
